query_optimizer/OptimizerTree.hpp

```diff
@@ -118,6 +122,169 @@ class OptimizerTree : public OptimizerTreeBase {
+        : key_(key), value_(value) {  }
+
+    std::string getName() const {
+      return "ProtoProperty";
+    }
+
+    void getFieldStringItems(std::vector<std::string> *inline_field_names,
+                             std::vector<std::string> *inline_field_values,
+                             std::vector<std::string> *non_container_child_field_names,
+                             std::vector<TreeNodeType> *non_container_child_fields,
+                             std::vector<std::string> *container_child_field_names,
+                             std::vector<std::vector<TreeNodeType>> *container_child_fields) const {
+      inline_field_names->push_back(std::string("Property"));
+      inline_field_values->push_back(key_);
+
+      inline_field_names->push_back(std::string("Value"));
+      inline_field_values->push_back(value_);
+    }
+
+   private:
+    std::string key_;
+    std::string value_;
+  };  // class OptimizerProtoRepresentation
+
+ public:
+  OptimizerProtoRepresentation() : properties_() { }
+
+  std::string getName() const {
+    return "ProtoDescription";
+  }
+
+  /**
+   * @brief Add a key-value type property to this proto representation.
+   * @details Internally, the instance creates another node to represent the
+   *          given values.
+   *
+   * @param key A string key.
+   * @param value A string value.
+   */
+  void addProperty(std::string key, std::string value) {
+    std::shared_ptr<const OptimizerProtoPropertyRepresentation> property(
+        new OptimizerProtoPropertyRepresentation(key, value));
+    properties_.push_back(property);
+  }
+
+  /**
+   * @brief Add a key-value type property to this proto representation.
+   * @details Internally, the instance creates another node to represent the
+   *          given values.
+   *
+   * @param key A string key.
+   * @param value An int value which is converted into a string.
+   */
+  void addProperty(std::string key, int value) {
+    std::shared_ptr<const OptimizerProtoPropertyRepresentation> property(
+        new OptimizerProtoPropertyRepresentation(key, std::to_string(value)));
+    properties_.push_back(property);
+  }
+
+  void getFieldStringItems(std::vector<std::string> *inline_field_names,
+                           std::vector<std::string> *inline_field_values,
+                           std::vector<std::string> *non_container_child_field_names,
+                           std::vector<TreeNodeType> *non_container_child_fields,
+                           std::vector<std::string> *container_child_field_names,
+                           std::vector<std::vector<TreeNodeType>> *container_child_fields) const {
+    for (auto property : properties_) {
+      non_container_child_field_names->push_back("Property");
+      non_container_child_fields->push_back(property);
+    }
+  }
+  // A list of managed properties.
+  std::vector<std::shared_ptr<const OptimizerProtoPropertyRepresentation> > properties_;
```

Comment:
After reading these comments and thinking about how the `OptimizerProtoRepresentation` class will be used, I realize that the vector of shared pointers in the class:
```
// A list of managed properties.
std::vector<std::shared_ptr<const OptimizerProtoPropertyRepresentation> > properties_;
```
should actually be a vector of `unique_ptrs`. This is because  `OptimizerProtoRepresentation`, the parent node in the tree representation of the Proto message will be shared across the Logical/physical CreateTable instances. Its children, the `OptimizerProtoPropertyRepresentation` will be managed by their parent, so, as long as there is a shared_ptr handling the shared parent, the children can be safely handled in unique ptrs owned by the parent.

Even better, I'll reuse the PtrVector class. Thank you @jianqiao for pointing this out.